### PR TITLE
LB-213: Add API usage examples to docs

### PR DESCRIPTION
**Pull request by eshansingh**
# Summary

* This is a…
    * [ ] Bug fix
    * [ ] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [x] Other
* **Describe this change in 1-2 sentences**:
This adds fully executable usage examples in Python 3 for the current API version. Current version includes examples for:

* [x] ~~`submit-listens`~~
* [x] ~~`get-listens`~~
* [x] ~~`set-latest-import`~~
* [x] ~~`get-latest-import`~~

# Problem

* JIRA ticket: https://tickets.metabrainz.org/projects/LB/issues/LB-213

# Solution

Each endpoint has an example file that's executable and demonstrates an example usecase for that particular endpoint.

# Action

Some aspects of how long the examples are, what additional details are needed, what parts are extraneous (for example, is the `verify_args` function useful for examples) etc.

### CODE REVIEW

**Comment by eshansingh** [comment]
The Travis build is always failing in the same way (`TestInfluxListenStore.test_import_listens` is failing), and apparently, if you restart them they fix themselves.

**Comment by eshansingh** [comment]
All of the example *code* is done, but I haven't added any detail to the documentation regarding those examples. I'll add whatever I feel is necessary right now, but I would love some feedback regarding it. 

Additionally, my `set-latest-import` commit passed Travis on the first try, which is odd.

File: docs/dev/api-usage.rst [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+Prerequisites
+#############
+
+All the examples assume you have a development version of the ListenBrainz
+server set up on :code:`localhost`. Remember to set :code:`DEBUG` to :py:`True`
+in the config. When in production, you can replace :code:`localhost` with
+:code:`api.listenbrainz.org` to use the real API. In order to use either
+one, you'll need a token. You can find it under :code:`ROOT/profile/`
+when signed in, with :code:`ROOT` being either :code:`localhost` or
+:code:`api.listenbrainz.org`.
```
Comment (paramsingh) [inline docs/dev/api-usage.rst]: We don't want to encourage users to use `api.listenbrainz.org` for anything other than sending API requests. We should change this to `listenbrainz.org`.

File: docs/dev/api-usage.rst [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+.. caution::
+
+   Use the token from the version you're testing with. You can't
```
Comment (paramsingh) [inline docs/dev/api-usage.rst]: Ideally, we wouldn't want to encourage people to test the API by submitting test data to production. I think we should remove this.

File: docs/dev/api_usage_examples/common.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+ROOT = '127.0.0.1'
+API_VERSION = 1
+TOKEN = '93fb3a42-9177-48a0-b07a-e8a7019218f6' # Replace with your dev token.
+USERNAME = 'naiveaiguy' # Replace with the username from which you got the token above.
+AUTH_HEADER = {
+    "Authorization": f"Token {TOKEN}"
+}
+
+class Track(dict):
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/common.py]: Because these are simple API usage examples, I'd prefer to keep boilerplate code to a minimum. Just using a dictionary instead of creating new classes is something we prefer in LB (and other MeB Python projects), so I'd suggest removing this class and just creating a dict listen like

```
EXAMPLE_PAYLOAD = {
    "listen_type": "single",
    "payload": [
        {
            "listened_at": 1486449409,
            "track_metadata": {
                "artist_name": "Kanye West",
                "track_name": "Fade",
                "release_name": "The Life of Pablo"
            }
        }
    ]
}
```

File: docs/dev/api_usage_examples/common.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        "release_name": "Whenever you need somebody"
+    })
+
+class ListenType(Enum):
+    Single = "single"
+    Playing = "playing_now"
+    Import = "import"
+
+def construct_url(resource_url):
+    return f"http://{ROOT}/{API_VERSION}/{resource_url}"
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/common.py]: Although LB runs on Python 3.6, we try to write 3.5 code as readthedocs doesn't support Python 3.6. You could use .format here, although, I would prefer explicit URLs in the `requests.post` and `requests.get` calls as they would make understanding the examples and API endpoints much easier.

File: docs/dev/api_usage_examples/get_latest_import.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+'''
+get_latest_import.py
+'''
+
+from common import AUTH_HEADER, USERNAME, construct_url
+
+import requests as req
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/get_latest_import.py]: I would prefer `import requests`, renaming seems unnecessary.

File: docs/dev/api_usage_examples/get_latest_import.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+'''
+get_latest_import.py
+'''
+
+from common import AUTH_HEADER, USERNAME, construct_url
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/get_latest_import.py]: Since these are examples, I'd prefer to keep these explicit. So that people can just look at the `submit_listen` code and understand it, without having to go into `common.py` to understand what the `AUTH_HEADER` is exactly.

File: docs/dev/api_usage_examples/get_latest_import.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        A ValueError if the JSON in the response is invalid.
+        An IndexError if the JSON is not structured as expected.
+    """
+    url = construct_url("latest-import")
+
+    params = {
+        "user_name": username
+    }
+
+    response = req.get(url, params=params, headers=AUTH_HEADER)
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/get_latest_import.py]: I think most of the code should be in this function.

Something like

```
response = requests.get(
   url,
   params={
       'user_name': username,
   },
   headers={
       'Authorization: 'Token <your token>',
   }
)
```

would be much clearer.

File: docs/dev/api_usage_examples/get_listens.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    # Thanks to requests, if any of these values are None,
+    # they're not going to be added in the first place.
+    params = {
+        "min_ts": min_ts,
+        "max_ts": max_ts,
+        "count": count
+    }
+
+    response = req.get(url, params=params, headers=AUTH_HEADER)
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/get_listens.py]: Same comment as the one for get_latest_import here.

File: docs/dev/api_usage_examples/set_latest_import.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        An HTTPError if there's a failure.
+        A ValueError if the JSON response is invalid.
+    """
+    url = construct_url("latest-import")
+
+    json_data = {
+        "ts": timestamp
+    }
+
+    response = req.post(url, json=json_data, headers=AUTH_HEADER)
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/set_latest_import.py]: Same here, `json_data` and `AUTH_HEADER` should be much more explicit.

File: docs/dev/api_usage_examples/submit_listens.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    json_data = {
+        "listen_type": listen_type.value,
+        "payload": payload
+    }
+
+    response = req.post(url, json=json_data, headers=AUTH_HEADER)
+
+    response.raise_for_status() # Will raise in case of a non-OK status
+
+    return response.json()
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/submit_listens.py]: I think an example of the data that is returned here would be good to add too.

File: docs/dev/api_usage_examples/common.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+common.py
+'''
+
+from enum import Enum
+from time import time
+from datetime import datetime
+
+ROOT = '127.0.0.1'
+API_VERSION = 1
+TOKEN = '93fb3a42-9177-48a0-b07a-e8a7019218f6' # Replace with your dev token.
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/common.py]: Sharing a token is not a good idea, should be replaced with some boilerplate.

File: docs/dev/api_usage_examples/common.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+USERNAME = 'naiveaiguy' # Replace with the username from which you got the token above.
+AUTH_HEADER = {
+    "Authorization": f"Token {TOKEN}"
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/common.py]: Use `.format` here. Python 3.6 fstrings cause problems when readthedocs builds the documentations.

File: docs/dev/api_usage_examples/get_listens.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    return response.json()['payload']['listens']
+
+if __name__ == "__main__":
+    listens = get_listens(USERNAME)
+    listens_tracks = [Track(track_dict) for track_dict in listens]
+
+    for track in listens_tracks:
+        print(track.simple_repr())
+
+def verify_args(username, min_ts=None, max_ts=None, count=None):
```
Comment (paramsingh) [inline docs/dev/api_usage_examples/get_listens.py]: This does not seem to be relevant to the api endpoints much. Maybe it can be removed.

**Comment by paramsingh** [comment]
@jflory7: I think a review from you would be helpful here too.

File: docs/dev/api-usage.rst (hunk 2 above)
Comment (eshansingh) [reply docs/dev/api-usage.rst]: Okay, I'll rewrite this to say "You should use the token from the API you're using. In production, change the token to one from `listenbrainz.org.`" Is that good?

File: docs/dev/api-usage.rst (hunk 1 above)
Comment (eshansingh) [reply docs/dev/api-usage.rst]: I don't quite get what you mean here, apologies. Do you mean for the token? Or in general? Because in production, AFAIK, you have to use `api.listenbrainz.org`, right?

File: docs/dev/api_usage_examples/common.py (hunk 12 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/common.py]: Sure.

File: docs/dev/api_usage_examples/common.py (hunk 3 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/common.py]: My reasoning for this was that it was more semantically understandable in examples. I'm not sure about just using a dictionary.

File: docs/dev/api_usage_examples/common.py (hunk 4 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/common.py]: I figured it would make it easier to update for users, and for us when the API version changes. Though to be fair, if the API version changes, we'll have to change the code anyway, so I'll just remove this function and use explicit URLs.

File: docs/dev/api_usage_examples/get_latest_import.py (hunk 6 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/get_latest_import.py]: `common.py` is fairly short, and it's given and the top of `api-usage.rst`. Should I just copy the required things that are in `common.py` right into the start of every file? That seems excessive to me, and it would only increase the size of the examples.

File: docs/dev/api_usage_examples/get_latest_import.py (hunk 5 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/get_latest_import.py]: Sorry, just a habit of mine. I'm a lazy Emacs user 😝

File: docs/dev/api_usage_examples/get_latest_import.py (hunk 7 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/get_latest_import.py]: Yeah, you're right.

File: docs/dev/api_usage_examples/get_listens.py (hunk 13 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/get_listens.py]: Yeah, even I wasn't sure about these - I'll remove both the `verify_args` functions.

File: docs/dev/api_usage_examples/submit_listens.py (hunk 10 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/submit_listens.py]: But this is `submit_listens.py`. If it all goes well, the only thing that's returned is `{"status": "ok"}`, which seems useless to include here.

**Comment by eshansingh** [comment]
Implemented some of the requested changes that I'm sure about. I'm still not sure what to do with the remains of `common.py`. If needed, I'll squash these small commits into a big refactoring commit once this PR is approved.

File: docs/dev/api_usage_examples/common.py (hunk 3 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/common.py]: Now that I think about it, both `Track` and `simple_repr` are only used in one file each. I'll simplify those places.

**Comment by eshansingh** [comment]
I've refactored `common.py` so it only includes constants. I've removed the `Track` class and the `ListenType` enum.

File: docs/dev/api_usage_examples/common.py (hunk 3 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/common.py]: Done.

File: docs/dev/api-usage.rst (hunk 1 above)
Comment (paramsingh) [reply docs/dev/api-usage.rst]: In general, you should get your production auth token from `listenbrainz.org` instead of `api.listenbrainz.org`. `listenbrainz.org` is the user facing site while `api.listenbrainz.org` is the place which handles the API calls.

File: docs/dev/api_usage_examples/common.py (hunk 11 above)
Comment (paramsingh) [reply docs/dev/api_usage_examples/common.py]: I know, but still feels weird to share ANY actual token.

File: docs/dev/api_usage_examples/get_latest_import.py (hunk 6 above)
Comment (paramsingh) [reply docs/dev/api_usage_examples/get_latest_import.py]: I think if we create examples that are easy to read individually without having to read the entire document, it would be better. That is why I suggested not using imports because then I have to go up to see what `AUTH_TOKEN` means, which is extra baggage if I'm a random guy who just wants to see how to submit listens and move on.

File: docs/dev/api_usage_examples/submit_listens.py (hunk 10 above)
Comment (paramsingh) [reply docs/dev/api_usage_examples/submit_listens.py]: That is something that we could mention in `api_usage.rst`.

"If everything worked, you should have received a json response of the form `{"status": "ok"}`"

File: docs/dev/api_usage_examples/common.py (hunk 11 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/common.py]: Changed it to `YOUR_TOKEN_HERE`.

File: docs/dev/api_usage_examples/submit_listens.py (hunk 10 above)
Comment (eshansingh) [reply docs/dev/api_usage_examples/submit_listens.py]: Ah, I see. Will do.

**Comment by eshansingh** [comment]
Implemented last few requested changes. Again, I'll be happy to `git rebase` the hell out of these commits if they're too many.

**Comment by eshansingh** [comment]
Squashed all of 'em.
